x/evmutil/keeper/grpc_query.go

```diff
@@ -33,3 +35,79 @@ func (s queryServer) Params(stdCtx context.Context, req *types.QueryParamsReques
 
 	return &types.QueryParamsResponse{Params: params}, nil
 }
+
+// DeployedCosmosCoinContracts gets contract addresses for deployed erc20 contracts
+// representing cosmos-sdk coins
+func (s queryServer) DeployedCosmosCoinContracts(
+	goCtx context.Context,
+	req *types.QueryDeployedCosmosCoinContractsRequest,
+) (res *types.QueryDeployedCosmosCoinContractsResponse, err error) {
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if len(req.CosmosDenoms) > 0 {
+		res, err = getDeployedCosmosCoinContractsByDenoms(&s.keeper, ctx, req.CosmosDenoms)
+	} else {
+		// requesting no sdk denoms is a request for all denoms
+		res, err = getAllDeployedCosmosCoinContractsPage(&s.keeper, ctx, req.Pagination)
+	}
+
+	return res, err
+}
+
+// getAllDeployedCosmosCoinContractsPage gets a page of deployed contracts (no filtering)
+func getAllDeployedCosmosCoinContractsPage(
+	k *Keeper, ctx sdk.Context, pagination *query.PageRequest,
+) (*types.QueryDeployedCosmosCoinContractsResponse, error) {
+	contracts := make([]types.DeployedCosmosCoinContract, 0)
+	contractStore := prefix.NewStore(
+		ctx.KVStore(k.storeKey),
+		types.DeployedCosmosCoinContractKeyPrefix,
+	)
+
+	pageRes, err := query.FilteredPaginate(contractStore, pagination,
+		func(key []byte, value []byte, accumulate bool) (bool, error) {
+			if !accumulate {
+				return true, nil
+			}
+			address := types.BytesToInternalEVMAddress(value)
+			contract := types.DeployedCosmosCoinContract{
+				CosmosDenom: string(key),
+				Address:     &address,
+			}
+			contracts = append(contracts, contract)
+			return true, nil
+		})
+	if err != nil {
+		return &types.QueryDeployedCosmosCoinContractsResponse{}, err
+	}
+
+	return &types.QueryDeployedCosmosCoinContractsResponse{
+		DeployedCosmosCoinContracts: contracts,
+		Pagination:                  pageRes,
+	}, nil
+}
+
+func getDeployedCosmosCoinContractsByDenoms(
+	k *Keeper, ctx sdk.Context, denoms []string,
+) (*types.QueryDeployedCosmosCoinContractsResponse, error) {
+	if len(denoms) > query.DefaultLimit {
```

Comment:
right, range & filter would allow pagination on this endpoint when requested with a >100 denoms query, too. but i made the tradeoff with the assumption that the most common use cases were either
1. looking up all contract addresses (no denoms in query)
2. looking up contract address for only a handful of contracts (limited number of denoms in query)

i think the majority of the time, 2) is going to be more efficient with direct lookups. a larger factor was that this split between a method that paginates all results & a method that only does direct lookups on a list felt easier to read in the code than the single PaginateFilter that handled both cases.
a nice bonus is that denoms that exist are returned in the order they were requested in (though omitting missing denoms mean there's no guarantee that search index == result index)